sdks/python/apache_beam/utils/windowed_value.py

```diff
@@ -178,33 +182,14 @@ def __repr__(self):
         self.windows,
         self.pane_info)
 
-  def __hash__(self):
-    return (hash(self.value) +
-            3 * self.timestamp_micros +
-            7 * hash(self.windows) +
-            11 * hash(self.pane_info))
-
-  # We'd rather implement __eq__, but Cython supports that via __richcmp__
-  # instead.  Fortunately __cmp__ is understood by both (but not by Python 3).
-  def __cmp__(left, right):  # pylint: disable=no-self-argument
-    """Compares left and right for equality.
-
-    For performance reasons, doesn't actually impose an ordering
-    on unequal values (always returning 1).
-    """
-    if type(left) is not type(right):
-      return cmp(type(left), type(right))
+  def _key(self):
+    return self.value, self.timestamp_micros, self.windows, self.pane_info
 
-    # TODO(robertwb): Avoid the type checks?
-    # Returns False (0) if equal, and True (1) if not.
-    return not WindowedValue._typed_eq(left, right)
+  def __eq__(self, other):
+    return type(self) == type(other) and self._key() == other._key()
```

Comment:
@robertwb should \_\_eq\_\_ be defined in the .pxd file or is this optimized automatically?

Reply:
__eq__ is a builtin (slots) method that is optimized automatically.